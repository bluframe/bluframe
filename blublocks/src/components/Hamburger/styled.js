// @flow

import styled, { type StyledComponent } from "styled-components"
import type { Theme } from "@bluframe/blublocks"

export const Bar: StyledComponent<{||}, Theme, HTMLSpanElement> = styled.span`
  background: ${({ theme }) => theme.palette.secondary.main};
  display: block;
  width: 24px;
  height: 4px;
  border-radius: 2px;

  + span {
    margin-top: 4px;
  }
`

export const Button: StyledComponent<
  {||},
  Theme,
  HTMLButtonElement
> = styled.button`
  position: relative;
  float: right;
  padding: 9px 10px;
  margin-top: 18px;
  margin-right: 0;
  margin-bottom: 8px;
  background-color: transparent;
  background-image: none;
  border: none;
  border-radius: 4px;

  &:focus {
    background-color: white;
    outline: none;
  }
`

export const ScreenReaderOnly: StyledComponent<
  any,
  mixed,
  HTMLSpanElement
> = styled.span`
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  border: 0;
`
