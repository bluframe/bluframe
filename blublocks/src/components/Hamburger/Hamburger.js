// @flow

import { Bar, Button, ScreenReaderOnly } from "./styled"
import type { ComponentProps } from "."
import React from "react"

const Hamburger = ({ onClick }: ComponentProps): React$Node => (
  <Button onClick={onClick} type="button">
    <ScreenReaderOnly>{"Toggle navigation"}</ScreenReaderOnly>

    <Bar />
    <Bar />
    <Bar />
  </Button>
)

export default Hamburger
