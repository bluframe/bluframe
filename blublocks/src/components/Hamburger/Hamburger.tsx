import { Bar, Button, ScreenReaderOnly } from "./styled"
import type { ComponentProps } from "."
import React from "react"

const Hamburger = ({
  onClick,
  isMenuExpanded = false,
  menuId
}: ComponentProps) => (
  <Button
    aria-controls={menuId}
    aria-expanded={isMenuExpanded}
    aria-label="Toggle navigation"
    onClick={onClick}
    role="menu"
    type="button"
  >
    <ScreenReaderOnly>{"Toggle navigation"}</ScreenReaderOnly>

    <Bar />
    <Bar />
    <Bar />
  </Button>
)

export default Hamburger
