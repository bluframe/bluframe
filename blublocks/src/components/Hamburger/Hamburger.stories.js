// @flow

import Hamburger, { type ComponentProps } from "components/Hamburger"
import React from "react"

const stories = {
  argTypes: {
    onClick: { action: "button click" }
  },
  component: Hamburger,
  title: "Components/Hamburger"
}

const Template = (args: ComponentProps) => <Hamburger {...args} />

export const Default: any = Template.bind({})

export default stories
