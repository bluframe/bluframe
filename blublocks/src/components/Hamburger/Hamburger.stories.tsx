import Hamburger, { Props } from "components/Hamburger"
import React from "react"

const stories = {
  argTypes: {
    onClick: { action: "button click" }
  },
  component: Hamburger,
  title: "Components/Hamburger"
}

const Template = (args: Props) => <Hamburger {...args} />

export const Default: any = Template.bind({})

export default stories
