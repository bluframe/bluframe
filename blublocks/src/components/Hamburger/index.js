// @flow

import Hamburger from "./Hamburger"
import type { HamburgerProps } from "@bluframe/blublocks"

export type ComponentProps = {|
  ...HamburgerProps
|}

export default Hamburger
