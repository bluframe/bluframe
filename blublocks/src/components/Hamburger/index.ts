import Hamburger from "./Hamburger"

export interface Props {
  isMenuExpanded?: boolean
  menuId?: string
  onClick?: () => void
}

export interface ComponentProps extends Props {}

export default Hamburger
