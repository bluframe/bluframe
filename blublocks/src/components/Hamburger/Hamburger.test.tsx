import { fireEvent, render } from "tests"
import Hamburger from "./Hamburger"
import React from "react"

const ONE = 1

const props = {
  isMenuExpanded: false,
  menuId: "main-navigation",
  onClick: jest.fn()
}

describe("Hamburger", () => {
  it("renders", () => {
    const { container } = render(<Hamburger {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("handles click event", () => {
    const { getByRole } = render(<Hamburger {...props} />)
    const button = getByRole("menu")

    fireEvent.click(button)

    expect(props.onClick).toHaveBeenCalledTimes(ONE)
  })

  it("sets aria-expanded attribute correctly", () => {
    const { getByRole, rerender } = render(<Hamburger {...props} />)
    const button = getByRole("menu")

    expect(button).toHaveAttribute("aria-expanded", "false")

    rerender(<Hamburger {...props} isMenuExpanded />)
    expect(button).toHaveAttribute("aria-expanded", "true")
  })

  it("sets aria-controls attribute correctly", () => {
    const { getByRole } = render(<Hamburger {...props} />)
    const button = getByRole("menu")

    expect(button).toHaveAttribute("aria-controls", "main-navigation")
  })
})
