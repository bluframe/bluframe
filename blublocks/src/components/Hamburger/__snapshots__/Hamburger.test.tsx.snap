// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hamburger renders 1`] = `
<button
  aria-controls="main-navigation"
  aria-expanded="false"
  aria-label="Toggle navigation"
  class="sc-gtcySc kGZfzc"
  role="menu"
  type="button"
>
  <span
    class="sc-dkXqGt czXpdt"
  >
    Toggle navigation
  </span>
  <span
    class="sc-bdDqHD dzqzmq"
  />
  <span
    class="sc-bdDqHD dzqzmq"
  />
  <span
    class="sc-bdDqHD dzqzmq"
  />
</button>
`;
